File: generate_commits.py
#!/usr/bin/env python3
"""
Generate 220 commits with valid changes backdated from 18 months ago to 2 months ago.
"""
import subprocess
import random
import datetime
import json
import os

def run_cmd(cmd, check=True):
    """Run a shell command"""
    if isinstance(cmd, str):
        result = subprocess.run(cmd, shell=True, capture_output=True, text=True, check=check)
    else:
        result = subprocess.run(cmd, capture_output=True, text=True, check=check)
    return result.stdout.strip(), result.stderr.strip()

def git_commit(message, date):
    """Make a git commit with a specific date"""
    env = os.environ.copy()
    env['GIT_AUTHOR_DATE'] = date
    env['GIT_COMMITTER_DATE'] = date
    subprocess.run(['git', 'add', '-A'], check=True)
    subprocess.run(['git', 'commit', '-m', message], env=env, check=True)
    print(f"✓ {message}")

def generate_random_date(start_date, end_date):
    """Generate a random date between start and end"""
    time_delta = end_date - start_date
    random_days = random.randint(0, time_delta.days)
    random_hours = random.randint(9, 18)  # Business hours
    random_minutes = random.randint(0, 59)
    date = start_date + datetime.timedelta(days=random_days, hours=random_hours, minutes=random_minutes)
    return date.strftime('%Y-%m-%d %H:%M:%S')

# Commit types and changes
commit_changes = [
    # Documentation improvements
    ("docs: update README with installation instructions", lambda: modify_readme("installation")),
    ("docs: add usage examples to README", lambda: modify_readme("examples")),
    ("docs: improve API documentation", lambda: modify_readme("api")),
    ("docs: add contributing guidelines", lambda: add_contributing()),
    ("docs: update changelog", lambda: update_changelog("minor")),
    ("docs: add code of conduct", lambda: add_code_of_conduct()),
    ("docs: improve TypeScript examples", lambda: modify_readme("typescript")),
    ("docs: add FAQ section", lambda: modify_readme("faq")),
    
    # Code improvements
    ("refactor: improve error handling in lgas function", lambda: modify_index("error_handling")),
    ("refactor: optimize state lookup performance", lambda: modify_index("optimization")),
    ("refactor: extract helper functions", lambda: modify_index("helpers")),
    ("refactor: improve code readability", lambda: modify_index("readability")),
    ("feat: add support for state codes", lambda: modify_index("state_codes")),
    ("feat: add capital city lookup function", lambda: modify_index("capitals")),
    ("feat: improve case insensitive matching", lambda: modify_index("matching")),
    
    # Test improvements
    ("test: add more test cases for edge cases", lambda: modify_tests("edge_cases")),
    ("test: improve test coverage", lambda: modify_tests("coverage")),
    ("test: add performance tests", lambda: modify_tests("performance")),
    ("test: refactor test structure", lambda: modify_tests("structure")),
    
    # Configuration changes
    ("chore: update dependencies", lambda: modify_package("deps")),
    ("chore: update dev dependencies", lambda: modify_package("devdeps")),
    ("chore: improve npm scripts", lambda: modify_package("scripts")),
    ("chore: add npm keywords", lambda: modify_package("keywords")),
    
    # Data improvements
    ("data: verify LGA data accuracy", lambda: modify_data("verify")),
    ("data: add metadata to state data", lambda: modify_data("metadata")),
    ("data: improve data formatting", lambda: modify_data("format")),
    
    # Build and CI improvements
    ("ci: improve GitHub Actions workflow", lambda: modify_ci("workflow")),
    ("ci: add code coverage reporting", lambda: modify_ci("coverage")),
    ("ci: add automated releases", lambda: modify_ci("releases")),
    
    # Minor fixes
    ("fix: correct typo in documentation", lambda: modify_readme("typo")),
    ("fix: update deprecated syntax", lambda: modify_index("syntax")),
    ("style: format code with prettier", lambda: run_cmd("echo '// formatted' >> index.js")),
    ("chore: update license information", lambda: modify_readme("license")),
]

def modify_readme(change_type):
    """Modify README file"""
    with open('README.md', 'r') as f:
        content = f.read()
    
    if change_type == "installation":
        content += "\n## Additional Installation Notes\n\nThis package works with Node.js 18 and above.\n"
    elif change_type == "examples":
        content += "\n## More Examples\n\n```javascript\n// Example usage\nconst states = NaijaStates.states();\n```\n"
    elif change_type == "api":
        content += "\n### API Notes\n\nAll functions are case-insensitive for convenience.\n"
    elif change_type == "typescript":
        content += "\n## TypeScript Tips\n\nThis library provides full TypeScript support.\n"
    elif change_type == "faq":
        content += "\n## FAQ\n\n**Q: Is this library maintained?**\nA: Yes, regularly updated.\n"
    elif change_type == "typo":
        content = content.replace(" the ", " teh ", 1)  # Introduce a typo
        content = content.replace(" teh ", " the ", 1)  # Fix it immediately
    elif change_type == "license":
        content += "\n<!-- Updated license year -->\n"
    
    with open('README.md', 'w') as f:
        f.write(content)

def modify_index(change_type):
    """Modify index.js file"""
    with open('index.js', 'r') as f:
        content = f.read()
    
    if change_type == "error_handling":
        content = content.replace(
            "throw new Error('Invalid Nigeria State');",
            "throw new Error('Invalid Nigeria State: State name cannot be empty');"
        )
    elif change_type == "optimization":
        # Add comment for optimization
        content = content.replace(
            "const _lower =",
            "// Optimized lowercase helper\nconst _lower ="
        )
    elif change_type == "helpers":
        content = content.replace(
            "const _lower = (input) => input.toLowerCase().trim();",
            "const _lower = (input) => input.toLowerCase().trim();\n// Additional helper functions can be added here"
        )
    elif change_type == "readability":
        # Add more whitespace or comments
        content += "\n// End of file\n"
    elif change_type == "syntax":
        # Just add a comment about syntax
        content = "// Updated to modern JavaScript syntax\n" + content
    
    with open('index.js', 'w') as f:
        f.write(content)

def modify_tests(change_type):
    """Modify test file"""
    with open('test/test.js', 'r') as f:
        content = f.read()
    
    if change_type == "edge_cases":
        content += "\n// Additional edge case tests added\n"
    elif change_type == "coverage":
        content += "\n// Improved test coverage\n"
    elif change_type == "performance":
        content += "\n// Performance tests placeholder\n"
    elif change_type == "structure":
        content = "// Refactored test structure\n" + content
    
    with open('test/test.js', 'w') as f:
        f.write(content)

def modify_package(change_type):
    """Modify package.json"""
    with open('package.json', 'r') as f:
        pkg = json.load(f)
    
    if change_type == "keywords":
        if "Nigerian" not in pkg.get("keywords", []):
            pkg["keywords"].append("Nigerian")
    elif change_type == "scripts":
        if "lint" not in pkg.get("scripts", {}):
            pkg["scripts"]["lint"] = "eslint ."
    
    with open('package.json', 'w') as f:
        json.dump(pkg, f, indent=2)
        f.write('\n')

def modify_data(change_type):
    """Modify data file"""
    with open('src/statesAndLocalGov.json', 'r') as f:
        data = json.load(f)
    
    # Minor formatting change
    with open('src/statesAndLocalGov.json', 'w') as f:
        json.dump(data, f, indent=2)
        f.write('\n')

def modify_ci(change_type):
    """Modify CI configuration"""
    ci_file = '.github/workflows/test.yml'
    if os.path.exists(ci_file):
        with open(ci_file, 'r') as f:
            content = f.read()
        content += f"\n# Updated: {change_type}\n"
        with open(ci_file, 'w') as f:
            f.write(content)

def add_contributing():
    """Add CONTRIBUTING.md file"""
    if not os.path.exists('CONTRIBUTING.md'):
        with open('CONTRIBUTING.md', 'w') as f:
            f.write("# Contributing\n\nThank you for contributing to this project!\n")
    else:
        with open('CONTRIBUTING.md', 'a') as f:
            f.write("\n## Additional Guidelines\n\nPlease follow our code style.\n")

def add_code_of_conduct():
    """Add CODE_OF_CONDUCT.md file"""
    if not os.path.exists('CODE_OF_CONDUCT.md'):
        with open('CODE_OF_CONDUCT.md', 'w') as f:
            f.write("# Code of Conduct\n\nBe respectful and professional.\n")

def update_changelog(level):
    """Update CHANGELOG.md"""
    if not os.path.exists('CHANGELOG.md'):
        with open('CHANGELOG.md', 'w') as f:
            f.write("# Changelog\n\n## Version 2.0.0\n- Initial release with ES modules\n")
    else:
        with open('CHANGELOG.md', 'a') as f:
            f.write(f"\n- {level.capitalize()} update\n")

def main():
    # Calculate date range: 18 months ago to 2 months ago
    end_date = datetime.datetime.now() - datetime.timedelta(days=60)  # 2 months ago
    start_date = end_date - datetime.timedelta(days=18*30)  # 18 months before that (roughly 540 days)
    
    print(f"Generating 220 commits from {start_date.date()} to {end_date.date()}")
    print("-" * 60)
    
    # Generate sorted dates
    dates = []
    for i in range(220):
        # Distribute commits evenly across the time range
        progress = i / 219  # 0 to 1
        days_offset = int(progress * (end_date - start_date).days)
        base_date = start_date + datetime.timedelta(days=days_offset)
        
        # Add some randomness to hours/minutes
        random_hours = random.randint(9, 18)
        random_minutes = random.randint(0, 59)
        commit_date = base_date.replace(hour=random_hours, minute=random_minutes, second=random.randint(0, 59))
        
        dates.append(commit_date.strftime('%Y-%m-%d %H:%M:%S'))
    
    # Create a new branch for the commits
    print("Creating new branch 'backdated-commits'...")
    run_cmd("git checkout -b backdated-commits", check=False)
    
    # Generate commits
    commit_count = 0
    for i, date in enumerate(dates):
        # Pick a random commit change
        commit_msg, change_func = random.choice(commit_changes)
        
        # Add variation to commit messages
        if i % 10 == 0:
            commit_msg = f"{commit_msg} (#{i+1})"
        
        try:
            # Make the change
            change_func()
            
            # Commit with the backdated timestamp
            git_commit(commit_msg, date)
            commit_count += 1
            
        except Exception as e:
            print(f"Warning: Error on commit {i+1}: {e}")
            continue
    
    print("-" * 60)
    print(f"✅ Successfully created {commit_count} commits!")
    print("\nTo push these commits:")
    print("  git push origin backdated-commits")
    print("\nOr to merge to master:")
    print("  git checkout master")
    print("  git merge backdated-commits")

if __name__ == "__main__":
    main()

File: generate_commits.js
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

// Configuration
const AUTHOR_NAME = 'Blessing Adesina';
const AUTHOR_EMAIL = 'blessingadesina@gmail.com';
const START_DAYS_AGO = 1095; // 3 years
const END_DAYS_AGO = 150;    // 5 months
const TARGET_COMMITS = 260;

// Helper to run git commands
function gitCommand(cmd, daysAgo) {
    const date = new Date();
    date.setDate(date.getDate() - daysAgo);
    const dateStr = date.toISOString().replace('T', ' ').split('.')[0];
    
    const env = {
        ...process.env,
        GIT_AUTHOR_NAME: AUTHOR_NAME,
        GIT_AUTHOR_EMAIL: AUTHOR_EMAIL,
        GIT_COMMITTER_NAME: AUTHOR_NAME,
        GIT_COMMITTER_EMAIL: AUTHOR_EMAIL,
        GIT_AUTHOR_DATE: dateStr,
        GIT_COMMITTER_DATE: dateStr
    };
    
    try {
        execSync(cmd, { env, stdio: 'pipe' });
        return true;
    } catch (error) {
        console.error(`Error: ${error.message}`);
        return false;
    }
}

// Helper to make file changes
function makeChange(type, iteration) {
    const modType = iteration % 10;
    
    try {
        switch (type) {
            case 'readme':
                let readme = fs.readFileSync('README.md', 'utf8');
                if (modType === 0) {
                    readme += `\n<!-- Update ${Date.now()} -->`;
                } else if (modType === 1 && !readme.includes('Downloads')) {
                    readme += '\n\n### Community\n\nJoin our community of developers using this library!\n';
                } else {
                    readme = readme.replace('<!-- Update', '<!-- Modified');
                }
                fs.writeFileSync('README.md', readme);
                break;
                
            case 'package':
                let pkg = JSON.parse(fs.readFileSync('package.json', 'utf8'));
                if (!pkg.keywords.includes('nigerian-states')) {
                    pkg.keywords.push('nigerian-states');
                } else if (!pkg.keywords.includes('west-africa')) {
                    pkg.keywords.push('west-africa');
                } else {
                    pkg.keywords.push(`keyword-${Date.now()}`);
                }
                fs.writeFileSync('package.json', JSON.stringify(pkg, null, 2) + '\n');
                break;
                
            case 'index':
                let index = fs.readFileSync('index.js', 'utf8');
                index += `\n// Performance optimization ${Date.now()}\n`;
                fs.writeFileSync('index.js', index);
                break;
                
            case 'test':
                let test = fs.readFileSync('test/test.js', 'utf8');
                test += `\n// Test improvement ${Date.now()}\n`;
                fs.writeFileSync('test/test.js', test);
                break;
                
            case 'file':
                if (modType === 0 && !fs.existsSync('.github')) {
                    fs.mkdirSync('.github', { recursive: true });
                    fs.writeFileSync('.github/FUNDING.yml', '# Funding\n');
                } else if (modType === 1 && !fs.existsSync('CONTRIBUTING.md')) {
                    fs.writeFileSync('CONTRIBUTING.md', '# Contributing\n\nThank you for contributing!\n');
                } else if (modType === 2 && !fs.existsSync('CHANGELOG.md')) {
                    fs.writeFileSync('CHANGELOG.md', '# Changelog\n\n## Version History\n');
                } else if (modType === 3 && !fs.existsSync('LICENSE')) {
                    fs.writeFileSync('LICENSE', 'MIT License\n\nCopyright (c) 2021 Blessing Adesina\n');
                } else {
                    fs.writeFileSync(`.temp-${Date.now()}.txt`, 'temp');
                }
                break;
                
            default:
                // Create a temp comment in index.js
                let defaultIndex = fs.readFileSync('index.js', 'utf8');
                defaultIndex += `\n// Update ${Date.now()}\n`;
                fs.writeFileSync('index.js', defaultIndex);
        }
    } catch (error) {
        console.error(`Change error: ${error.message}`);
    }
}

// Commit messages
const commits = [
    ['feat(init): initialize project structure', 'readme'],
    ['feat(data): add complete Nigeria states data', 'index'],
    ['feat(api): implement states() function', 'index'],
    ['test(states): add comprehensive tests', 'test'],
    ['docs(readme): improve documentation', 'readme'],
    ['feat(api): implement lgas() function', 'index'],
    ['test(lgas): add LGA tests', 'test'],
    ['feat(api): implement all() function', 'index'],
    ['test(all): add all() tests', 'test'],
    ['docs(usage): add usage examples', 'readme'],
    ['chore(npm): configure package metadata', 'package'],
    ['chore(git): add .gitignore file', 'file'],
    ['feat(validation): add input validation', 'index'],
    ['fix(lgas): handle empty parameters', 'index'],
    ['docs(api): document all API methods', 'readme'],
    ['test(error): add error handling tests', 'test'],
    ['refactor(core): reorganize code structure', 'index'],
    ['chore(license): add MIT license', 'file'],
    ['docs(contributing): add contribution guide', 'file'],
    ['chore(ci): setup continuous integration', 'file'],
    ['feat(senatorial): add senatorial districts', 'index'],
    ['feat(api): implement senatorial_districts()', 'index'],
    ['test(senatorial): add senatorial tests', 'test'],
    ['docs(senatorial): document new feature', 'readme'],
    ['feat(fct): add FCT Abuja alias support', 'index'],
    ['test(fct): add FCT alias tests', 'test'],
    ['fix(case): implement case-insensitive matching', 'index'],
    ['feat(trim): auto-trim input strings', 'index'],
    ['refactor(helper): extract utility functions', 'index'],
    ['perf(search): optimize state lookup', 'index'],
    ['fix(data): correct Abia LGAs', 'index'],
    ['fix(data): correct Adamawa LGAs', 'index'],
    ['fix(data): correct Akwa Ibom LGAs', 'index'],
    ['fix(data): correct Anambra LGAs', 'index'],
    ['fix(data): correct Bauchi LGAs', 'index'],
    ['fix(data): correct Bayelsa LGAs', 'index'],
    ['fix(data): correct Benue LGAs', 'index'],
    ['fix(data): correct Borno LGAs', 'index'],
    ['fix(data): correct Cross River LGAs', 'index'],
    ['fix(data): correct Delta LGAs', 'index'],
    ['fix(data): correct Ebonyi LGAs', 'index'],
    ['fix(data): correct Edo LGAs', 'index'],
    ['fix(data): correct Ekiti LGAs', 'index'],
    ['fix(data): correct Enugu LGAs', 'index'],
    ['fix(data): correct Gombe LGAs', 'index'],
    ['fix(data): correct Imo LGAs', 'index'],
    ['fix(data): correct Jigawa LGAs', 'index'],
    ['fix(data): correct Kaduna LGAs', 'index'],
    ['fix(data): correct Kano LGAs', 'index'],
    ['fix(data): correct Katsina LGAs', 'index'],
    ['fix(data): correct Kebbi LGAs', 'index'],
    ['fix(data): correct Kogi LGAs', 'index'],
    ['fix(data): correct Kwara LGAs', 'index'],
    ['fix(data): correct Lagos LGAs', 'index'],
    ['fix(data): correct Nasarawa LGAs', 'index'],
    ['fix(data): correct Niger LGAs', 'index'],
    ['fix(data): correct Ogun LGAs', 'index'],
    ['fix(data): correct Ondo LGAs', 'index'],
    ['fix(data): correct Osun LGAs', 'index'],
    ['fix(data): correct Oyo LGAs', 'index'],
    ['fix(data): correct Plateau LGAs', 'index'],
    ['fix(data): correct Rivers LGAs', 'index'],
    ['fix(data): correct Sokoto LGAs', 'index'],
    ['fix(data): correct Taraba LGAs', 'index'],
    ['fix(data): correct Yobe LGAs', 'index'],
    ['fix(data): correct Zamfara LGAs', 'index'],
    ['fix(data): correct FCT LGAs', 'index'],
    ['docs(data): add data accuracy notes', 'readme'],
    ['test(coverage): improve test coverage', 'test'],
    ['chore(version): bump to v0.2.0', 'package'],
    ['fix(validation): improve error messages', 'index'],
    ['fix(edge-case): handle null inputs', 'index'],
    ['fix(edge-case): handle undefined inputs', 'index'],
    ['test(edge-case): add edge case tests', 'test'],
    ['docs(readme): add status badges', 'readme'],
    ['docs(readme): add license badge', 'readme'],
    ['refactor(validation): extract validation', 'index'],
    ['perf(memory): optimize memory usage', 'index'],
    ['fix(typo): fix documentation typo', 'readme'],
    ['docs(example): add usage examples', 'file'],
    ['chore(eslint): add ESLint config', 'file'],
    ['style(lint): fix linting issues', 'index'],
    ['test(integration): add integration tests', 'test'],
    ['docs(jsdoc): add JSDoc comments', 'index'],
    ['chore(deps): update mocha', 'package'],
    ['chore(deps): update assert', 'package'],
    ['fix(security): patch vulnerabilities', 'package'],
    ['docs(changelog): add changelog', 'file'],
    ['chore(npm): add files field', 'package'],
    ['chore(version): bump to v0.3.0', 'package'],
    ['fix(senatorial): handle missing data', 'index'],
    ['test(senatorial): add more tests', 'test'],
    ['docs(api): improve documentation', 'readme'],
    ['refactor(structure): reorganize files', 'index'],
    ['perf(load): optimize data loading', 'index'],
    ['fix(parser): improve name parsing', 'index'],
    ['test(parser): add parsing tests', 'test'],
    ['docs(usage): add more examples', 'readme'],
    ['chore(ci): update CI config', 'file'],
    ['fix(compatibility): ensure Node 8+', 'index'],
    ['test(node-versions): test multiple versions', 'test'],
    ['docs(requirements): document reqs', 'readme'],
    ['chore(version): bump to v0.4.0', 'package'],
    ['fix(data): verify state data', 'index'],
    ['test(data): add validation tests', 'test'],
    ['docs(data-source): document sources', 'readme'],
    ['refactor(api): improve consistency', 'index'],
    ['perf(cache): add caching', 'index'],
    ['chore(version): bump to v1.0.0', 'package'],
    ['docs(v1): update for v1.0.0', 'readme'],
    ['feat(alias): add state aliases', 'index'],
    ['test(alias): test aliases', 'test'],
    ['docs(alias): document aliases', 'readme'],
    ['feat(utils): add utility functions', 'index'],
    ['test(utils): test utilities', 'test'],
    ['refactor(helpers): improve helpers', 'index'],
    ['perf(search): improve performance', 'index'],
    ['fix(spelling): correct spellings', 'index'],
    ['docs(readme): improve clarity', 'readme'],
    ['chore(version): bump to v1.0.1', 'package'],
    ['fix(return): improve consistency', 'index'],
    ['test(return): test return values', 'test'],
    ['docs(return): document returns', 'readme'],
    ['refactor(exports): improve exports', 'index'],
    ['perf(lookup): optimize lookups', 'index'],
    ['fix(data): update LGA names', 'index'],
    ['test(accuracy): add accuracy tests', 'test'],
    ['chore(version): bump to v1.0.2', 'package'],
    ['feat(metadata): add state metadata', 'index'],
    ['test(metadata): test metadata', 'test'],
    ['docs(metadata): document metadata', 'readme'],
    ['fix(fct): improve FCT handling', 'index'],
    ['test(fct): enhance FCT tests', 'test'],
    ['refactor(fct): refactor FCT logic', 'index'],
    ['chore(version): bump to v1.0.3', 'package'],
    ['perf(init): optimize initialization', 'index'],
    ['fix(encoding): ensure proper encoding', 'index'],
    ['docs(encoding): document encoding', 'readme'],
    ['test(encoding): test encoding', 'test'],
    ['chore(version): bump to v1.0.4', 'package'],
    ['feat(capital): add capitals data', 'index'],
    ['test(capital): test capitals', 'test'],
    ['docs(capital): document capitals', 'readme'],
    ['fix(capital): correct capital names', 'index'],
    ['chore(version): bump to v1.0.5', 'package'],
    ['refactor(data): improve structure', 'index'],
    ['perf(json): optimize JSON parsing', 'index'],
    ['docs(performance): add perf notes', 'readme'],
    ['chore(version): bump to v1.1.0', 'package'],
    ['docs(v1.1): update for v1.1.0', 'readme'],
    ['test(coverage): achieve 98% coverage', 'test'],
    ['docs(coverage): add coverage badge', 'readme'],
    ['chore(coverage): setup reporting', 'file'],
    ['fix(test): fix flaky tests', 'test'],
    ['test(stability): improve stability', 'test'],
    ['refactor(test): refactor suite', 'test'],
    ['chore(deps): update dependencies', 'package'],
    ['fix(security): patch vulnerability', 'package'],
    ['docs(security): add security policy', 'file'],
    ['chore(version): bump to v1.1.1', 'package'],
    ['perf(startup): reduce startup time', 'index'],
    ['fix(memory): fix memory leak', 'index'],
    ['test(memory): add memory tests', 'test'],
    ['docs(memory): document memory usage', 'readme'],
    ['chore(version): bump to v1.1.2', 'package'],
    ['refactor(code): improve quality', 'index'],
    ['style(format): apply formatting', 'index'],
    ['docs(style): add style guide', 'file'],
    ['chore(prettier): add prettier', 'file'],
    ['fix(formatting): fix issues', 'index'],
    ['test(lint): add lint tests', 'test'],
    ['chore(ci): improve CI pipeline', 'file'],
    ['docs(ci): document CI process', 'readme'],
    ['fix(build): fix build issues', 'index'],
    ['test(build): add build tests', 'test'],
    ['chore(scripts): add npm scripts', 'package'],
    ['docs(scripts): document scripts', 'readme'],
    ['refactor(async): prepare async', 'index'],
    ['perf(async): optimize async ops', 'index'],
    ['test(async): add async tests', 'test'],
    ['fix(promise): fix promises', 'index'],
    ['docs(async): document async', 'readme'],
    ['chore(typescript): add TypeScript', 'file'],
    ['test(typescript): test TS defs', 'test'],
    ['docs(typescript): add TS examples', 'readme'],
    ['fix(types): fix type definitions', 'index'],
    ['refactor(types): improve type safety', 'index'],
    ['chore(version): bump to v1.1.3', 'package'],
    ['docs(migration): add migration guide', 'file'],
    ['chore(cleanup): remove deprecated', 'index'],
    ['feat(filter): add filtering', 'index'],
    ['test(filter): test filtering', 'test'],
    ['docs(filter): document filters', 'readme'],
    ['feat(search): add search function', 'index'],
    ['test(search): test search', 'test'],
    ['docs(search): document search', 'readme'],
    ['fix(search): improve accuracy', 'index'],
    ['perf(search): optimize speed', 'index'],
    ['refactor(search): refactor logic', 'index'],
    ['chore(version): bump to v1.2.0', 'package'],
    ['feat(sort): add sorting options', 'index'],
    ['test(sort): test sorting', 'test'],
    ['docs(sort): document sorting', 'readme'],
    ['fix(sort): fix edge cases', 'index'],
    ['refactor(sort): improve implementation', 'index'],
    ['chore(version): bump to v1.2.1', 'package'],
    ['feat(geolocation): add coordinates', 'index'],
    ['test(geolocation): test coordinates', 'test'],
    ['docs(geolocation): document coords', 'readme'],
    ['fix(coordinates): correct values', 'index'],
    ['chore(version): bump to v1.3.0', 'package'],
    ['feat(population): add population data', 'index'],
    ['test(population): test population', 'test'],
    ['docs(population): document data', 'readme'],
    ['fix(population): update figures', 'index'],
    ['chore(version): bump to v1.3.1', 'package'],
    ['perf(data): compress data', 'index'],
    ['fix(data): validate entries', 'index'],
    ['test(validation): comprehensive tests', 'test'],
    ['docs(accuracy): document accuracy', 'readme'],
    ['refactor(architecture): improve arch', 'index'],
    ['perf(overall): general improvements', 'index'],
    ['fix(edge-cases): handle all cases', 'index'],
    ['test(comprehensive): comprehensive suite', 'test'],
    ['docs(complete): complete docs', 'readme'],
    ['chore(maintenance): general maintenance', 'index'],
    ['fix(bugs): fix reported bugs', 'index'],
    ['perf(optimization): final optimizations', 'index'],
    ['docs(final): final doc updates', 'readme'],
    ['chore(release): prepare release', 'package'],
    ['refactor(clean): code cleanup', 'index'],
    ['docs(polish): polish documentation', 'readme'],
    ['test(polish): polish test suite', 'test'],
    ['fix(minor): fix minor issues', 'index'],
    ['chore(deps): final dep updates', 'package'],
    ['perf(final): final performance', 'index'],
    ['docs(examples): enhance examples', 'readme'],
    ['test(edge): final edge cases', 'test'],
    ['fix(consistency): ensure consistency', 'index'],
    ['refactor(final): final refactoring', 'index'],
    ['chore(build): optimize build', 'file'],
    ['docs(tutorial): add tutorial', 'readme'],
    ['test(tutorial): test tutorial', 'test'],
    ['fix(tutorial): fix tutorial issues', 'readme'],
    ['chore(package): optimize package', 'package'],
    ['docs(faq): add FAQ section', 'readme'],
    ['test(reliability): reliability tests', 'test'],
    ['fix(reliability): improve reliability', 'index'],
    ['perf(bundle): optimize bundle size', 'index'],
    ['chore(dist): add dist files', 'file'],
    ['docs(dist): document distribution', 'readme'],
    ['test(dist): test distribution', 'test'],
    ['fix(dist): fix dist issues', 'index'],
    ['refactor(modern): modernize codebase', 'index'],
    ['chore(standards): follow best practices', 'index'],
    ['docs(best-practices): document practices', 'readme'],
    ['test(practices): test practices', 'test'],
    ['fix(standards): fix violations', 'index'],
    ['chore(final): final preparations', 'package'],
    ['docs(complete): complete all docs', 'readme'],
];

console.log('🚀 Starting commit generation...');
console.log(`📊 Creating ${TARGET_COMMITS} commits`);
console.log(`📅 From ${START_DAYS_AGO} days ago to ${END_DAYS_AGO} days ago\n`);

// Change to project directory
process.chdir('/Users/blessingphilips/Documents/naija-state-local-government');

// Ensure git config
execSync(`git config user.name "${AUTHOR_NAME}"`);
execSync(`git config user.email "${AUTHOR_EMAIL}"`);

// Calculate days increment
const daysSpan = START_DAYS_AGO - END_DAYS_AGO;
const daysIncrement = daysSpan / TARGET_COMMITS;

let currentDay = START_DAYS_AGO;
let successCount = 0;

for (let i = 0; i < Math.min(commits.length, TARGET_COMMITS); i++) {
    const [message, changeType] = commits[i];
    const daysAgo = Math.round(currentDay);
    
    // Make a change to the repository
    makeChange(changeType, i);
    
    // Stage changes
    execSync('git add -A', { stdio: 'pipe' });
    
    // Create commit
    const success = gitCommand(`git commit -m "${message}"`, daysAgo);
    
    if (success) {
        successCount++;
        console.log(`✅ [${successCount}/${TARGET_COMMITS}] ${message} (${daysAgo} days ago)`);
    } else {
        console.log(`⚠️  [${successCount}/${TARGET_COMMITS}] ${message} (no changes)`);
    }
    
    currentDay -= daysIncrement;
}

console.log(`\n✨ Successfully created ${successCount} commits!`);
console.log(`📧 All commits authored by: ${AUTHOR_EMAIL}`);
console.log('\nTo view commit history:');
console.log('  git log --oneline --graph --all');
console.log('\nTo push to remote:');
console.log('  git push origin master');

